packages/gatsby-image/src/index.js

```diff
@@ -74,9 +74,9 @@ const matchesMedia = ({ media }) =>
  * @return {string}
  */
 const getImageSrcKey = ({ fluid, fixed }) => {
-  const data = fluid ? getCurrentSrcData(fluid) : getCurrentSrcData(fixed)
+  const currentData = fluid || fixed
 
-  return data.src
+  return currentData && getCurrentSrcData(currentData).src
```

Comment:
Is this where the error is actually occurring, or within the function call logic itself?

I see `getCurrentSrcData()` use a conditional, with the 2nd condition being a method that runs checks, those would abort and not trigger further errors, except for the end of the function, where it's assumed an array has been provided to the function. The other method callers are in the fluid/fixed condition checks in render method, they should be fine.

The error here is from trying to access the `src` property? Your fix will return false otherwise, where `getImageSrcKey()` would be affected when creating a cache entry under `false` key. This can apparently be a valid key for an object, which may cause a problem if several `<Img />` utilize the same cache key?(I'm not sure if the cache object is shared among instances)

https://github.com/gatsbyjs/gatsby/blob/6102456c4a9f0a09aad5e3c2d61d6a38591ded2d/packages/gatsby-image/src/index.js#L120

If that would be an issue we'd want to address that.

---

I think this is more clear for where guard is appropriate:

```js
  const currentData = getCurrentSrcData(fluid || fixed)
  return currentData && currentData.src
```

Probably still want to avoid caching `false` as a cache key though if that's returned instead.

---

https://github.com/gatsbyjs/gatsby/blob/6102456c4a9f0a09aad5e3c2d61d6a38591ded2d/packages/gatsby-image/src/index.js#L103-L104

`hasArtDirectionSupport()` is only called by one method atm, the method that ends with the above under the assumption there is an array of objects provided.

https://github.com/gatsbyjs/gatsby/blob/6102456c4a9f0a09aad5e3c2d61d6a38591ded2d/packages/gatsby-image/src/index.js#L58-L59

You could add this to the start of `getCurrentSrcData()` to abort early with false bool.

---

Alternatively, is this something that the component should actually support? (No fluid or fixed object provided) Or should an error be thrown here(or earlier) instead? (albeit with a friendlier message)

Reply:
@polarathene I agree a warning would be nicer! I hesitated to add one because I didn't want to bloat the client code. But I see now there's another warning in the image plugin that's gated by `process.env.NODE_ENV !== "production"`. I presume webpack removes that in production bundles. I'll do the same in this case then!

The error was `getCurrentSrcData(null)` crashing, but I didn't feel it reasonable that it should be able to handle being passed `null`. 

By the way, `null && whatever` evaluates to `null`, not `false`. You're right it's weird that `null` is later used as a cache key but I don't think it should cause any harm. 

Should i surround `imageCache[src] = true;` with an `if (src != null)` in `activateCacheForImage`? That way it's clear it's a non-cacheable value.

For better readability we could replace `null` with a constant `const invalidSrcData = null`, but that depends on your coding style.

Reply:
> The error was getCurrentSrcData(null) crashing, but I didn't feel it reasonable that it should be able to handle being passed null. And you would need to check for whether src is present if it does.

The stackoverflow error suggested it tried to access `currentData.0` as a value to return as no array was provided to index. The method is meant to return an object, not just for the `src` property that this particular bug clearly has no access to.

> Should i surround `imageCache[src] = true;` with an `if (src != null)` in `activateCacheForImage`? That way it's clear it's a non-cacheable value.

Yes, that'd probably be best, but just `if (src) { ... }` should be fine, while unlikely to be an issue, it guards against all falsy values, whereas `src != null` only covers `null` and `undefined`, plus it's more terse so win-win :)

> but that depends on your coding style.

Personally I prefer the `&&` short-circuit approach you've used. It's already used quite heavily in the file, so any contributors hopefully know that a [falsy value](https://developer.mozilla.org/en-US/docs/Glossary/Falsy) can be returned. 

There's also some [double bang(NOT)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_NOT#Examples) operator usage, in one case at least it's to ensure a boolean value is provided for IE as it couldn't handle `undefined` properly if I remember right.

Reply:
> @polarathene I agree a warning would be nicer! I hesitated to add one because I didn't want to bloat the client code. But I see now there's another warning in the image plugin that's gated by process.env.NODE_ENV !== "production". I presume webpack removes that in production bundles. I'll do the same in this case then!
> 
> About throwing an error instead: I suppose it's conceivable that this happens at runtime (after deployment) if you pass a variable that has a dynamic, calculated value. At runtime you wouldn't want to crash. But that's kind of a code standard thing and I don't know how you do error handling elsewhere. And maybe it's far-fetched. If that's not a consideration, halting with an error is better.

Yeah, let's make sure to log a warning here. Let's not throw an error but just log one. So something like:

```js
if (process.env.NODE_ENV !== `production`) {
  if (!currentData) {
   console.error(`gatsby-image requires a valid src property. (maybe more actionable info here..)`)
  }
}
```

Reply:
I wonder if we could just wrap this in a `try/catch`

Reply:
A `try/catch` would be good, perhaps with the following error message:

```
gatsby-image expected to receive a valid 'fixed' or 'fluid' prop, missing 'src' value.
```

If there isn't a valid reason for not supplying the image data prop(fixed / fluid), then the error could probably be fired off in the constructor. It's not going to do SSR any good either I imagine, the constructor will fire off on initial load too where adding some props will have no affect. 

If there's no intention to render the component initially, shouldn't they instead use conditional rendering? `{someCondition && <Img fluid={ ... } />}`.